### PR TITLE
Move MITRE reference with `external_id` to the exterior level of the response for all MITRE endpoints

### LINKED ISSUE #8879: MITRE endpoints can't sort by `references.external_id`
|Wazuh version|Component|
|---|---|
| 4.3.x | API |

| Related to | #7819 |
|---|---|

**Description**

The Wazuh Kibana App needs to sort the MITRE endpoints by the `external_id` field that is included in the references.
This field that is important for the users is the `public` ID for the MITRE resources.
Maybe it should be included in the main object.

image
image

### PR DESCRIPTION
**Pull request by mcarmona99**
|Related issue|
|---|
| #8879 |

Hi team

This PR closes #8879.

In this pull request, I have included the reference with `external_id` out of the references list of each MITRE item.

I have updated unittest and API integration tests to include this change.

`GET /mitre/mitigations?wait_for_complete=true&limit=1`

<details>

<summary>Response</summary>

```
{
  "data": {
    "affected_items": [
      {
        "deprecated": 1,
        "created_time": "2018-10-17 00:14:20.652000",
        "name": "Password Filter DLL Mitigation",
        "mitre_version": "1.0",
        "id": "course-of-action--00d7d21b-69d6-4797-88a2-c86f3fc97651",
        "description": "Ensure only valid password filters are registered. Filter DLLs must be present in Windows installation directory (<code>C:\\Windows\\System32\\</code> by default) of a domain controller and/or local computer with a corresponding entry in <code>HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Control\\Lsa\\Notification Packages</code>. (Citation: Microsoft Install Password Filter n.d)",
        "modified_time": "2019-07-25 11:22:19.139000",
        "techniques": [
          "attack-pattern--b8c5c9dd-a662-479d-9428-ae745872537c"
        ],
        "references": [
          {
            "source": "Microsoft Install Password Filter n.d",
            "url": "https://msdn.microsoft.com/library/windows/desktop/ms721766.aspx",
            "description": "Microsoft. (n.d.). Installing and Registering a Password Filter DLL. Retrieved November 21, 2017."
          }
        ],
        "source": "mitre-attack",
        "external_id": "T1174",
        "url": "https://attack.mitre.org/mitigations/T1174"
      }
    ],
    "total_affected_items": 266,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "MITRE mitigations information was returned",
  "error": 0
}
```

</details>

With these changes, we can now sort by external_id, source and url:

`GET /mitre/mitigations?wait_for_complete=true&limit=3&sort=-external_id&select=external_id,name
`

<details>

<summary>Response</summary>

```
{
  "data": {
    "affected_items": [
      {
        "external_id": "T1501",
        "name": "Systemd Service Mitigation",
        "id": "course-of-action--83130e62-bca6-4a81-bd4b-8e233bd49db6"
      },
      {
        "external_id": "T1500",
        "name": "Compile After Delivery Mitigation",
        "id": "course-of-action--ae56a49d-5281-45c5-ab95-70a1439c338e"
      },
      {
        "external_id": "T1499",
        "name": "Endpoint Denial of Service Mitigation",
        "id": "course-of-action--82c21600-ccb6-4232-8c04-ef3792b56628"
      }
    ],
    "total_affected_items": 266,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "MITRE mitigations information was returned",
  "error": 0
}
```

</details>

### Test results:

```
pytest framework/wazuh/core/tests/test_mitre.py 
=================== 13 passed, 2 warnings in 0.17s ===================

pytest framework/wazuh/tests/test_mitre.py 
=================== 7 passed, 2 warnings in 0.19s ===================
```

```
test_mitre_endpoints.tavern.yaml 
	 7 passed, 13 warnings

test_rbac_black_mitre_endpoints.tavern.yaml 
	 7 passed, 13 warnings

test_rbac_white_mitre_endpoints.tavern.yaml 
	 7 passed, 13 warnings
```

Regards,
Manuel.

### CODE REVIEW

File: framework/wazuh/core/mitre.py [hunk 1]
```diff
@@ -186,6 +190,17 @@ def _execute_data_query(self):
...
                                         row['id'] == mitigation['id']]
 
+            # Take reference with external_id checking it is not None
+            reference_external_id = next(
+                (row_no_id for row_no_id in mitigation['references'] if
+                 'external_id' in row_no_id and row_no_id['external_id']), {})
+            if 'description' in reference_external_id:
+                reference_external_id.pop('description')
+            # Delete the reference from references and update the object
+            if reference_external_id in mitigation['references']:
```
Comment (CarlosRS9) [inline framework/wazuh/core/mitre.py:200]:
```suggestion
            if reference_external_id:
```
If the `reference_external_id` is not empty it is guaranteed to be in `mitigation['references']`

File: framework/wazuh/core/mitre.py [hunk 2]
```diff
@@ -270,6 +286,17 @@ def _execute_data_query(self):
...
                                     row['id'] == tactic['id']]
 
+            # Take reference with external_id checking it is not None
+            reference_external_id = next(
+                (row_no_id for row_no_id in tactic['references'] if
+                 'external_id' in row_no_id and row_no_id['external_id']), {})
+            if 'description' in reference_external_id:
+                reference_external_id.pop('description')
+            # Delete the reference from references and update the object
+            if reference_external_id in tactic['references']:
```
Comment (CarlosRS9) [inline framework/wazuh/core/mitre.py:296]:
```suggestion
            if reference_external_id:
```

File: framework/wazuh/core/mitre.py [hunk 3]
```diff
@@ -346,6 +373,17 @@ def _execute_data_query(self):
...
                                        row['id'] == technique['id']]
 
+            # Take reference with external_id checking it is not None
+            reference_external_id = next(
+                (row_no_id for row_no_id in technique['references'] if
+                 'external_id' in row_no_id and row_no_id['external_id']), {})
+            if 'description' in reference_external_id:
+                reference_external_id.pop('description')
+            # Delete the reference from references and update the object
+            if reference_external_id in technique['references']:
```
Comment (CarlosRS9) [inline framework/wazuh/core/mitre.py:383]:
```suggestion
            if reference_external_id:
```

File: framework/wazuh/core/mitre.py [hunk 4]
```diff
@@ -411,6 +450,17 @@ def _execute_data_query(self):
...
                                    row['id'] == group['id']]
 
+            # Take reference with external_id checking it is not None
+            reference_external_id = next(
+                (row_no_id for row_no_id in group['references'] if
+                 'external_id' in row_no_id and row_no_id['external_id']), {})
+            if 'description' in reference_external_id:
+                reference_external_id.pop('description')
+            # Delete the reference from references and update the object
+            if reference_external_id in group['references']:
```
Comment (CarlosRS9) [inline framework/wazuh/core/mitre.py:460]:
```suggestion
            if reference_external_id:
```

File: framework/wazuh/core/mitre.py [hunk 5]
```diff
@@ -478,6 +529,17 @@ def _execute_data_query(self):
...
                                       row['id'] == software['id']]
 
+            # Take reference with external_id checking it is not None
+            reference_external_id = next(
+                (row_no_id for row_no_id in software['references'] if
+                 'external_id' in row_no_id and row_no_id['external_id']), {})
+            if 'description' in reference_external_id:
+                reference_external_id.pop('description')
+            # Delete the reference from references and update the object
+            if reference_external_id in software['references']:
```
Comment (CarlosRS9) [inline framework/wazuh/core/mitre.py:539]:
```suggestion
            if reference_external_id:
```

**Review by CarlosRS9 — DISMISSED** [review]
LGTM!

**Review by davidjiglesias — CHANGES_REQUESTED** [review]
Add new references handling to father class

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**